sympy/calculus/tests/test_util.py

```diff
@@ -31,3 +33,152 @@ def test_not_empty_in():
+    assert (-oo - Limits(-1, oo)) == -oo
+    assert Limits(-oo, 1) - oo == -oo
+
+    assert Limits(1, oo) - oo == Limits(-oo, oo)
+    assert Limits(-oo, 1) - (-oo) == Limits(-oo, oo)
+    assert (oo - Limits(1, oo)) == Limits(-oo, oo)
+    assert (-oo - Limits(1, oo)) == -oo
+
+    assert Limits(1, 2)/2 == Limits(S(1)/2, 1)
+    assert 2/Limits(2, 3) == Limits(S(2)/3, 1)
+    assert 1/Limits(-1, 1) == Limits(-oo, oo)
+
+    assert abs(Limits(1, 2)) == Limits(1, 2)
+    assert abs(Limits(-2, -1)) == Limits(1, 2)
+    assert abs(Limits(-2, 1)) == Limits(0, 2)
+    assert abs(Limits(-1, 2)) == Limits(0, 2)
+
+
+def test_Limits_mul():
+    assert Limits(1, 2)*2 == Limits(2, 4)
+    assert 2*Limits(1, 2) == Limits(2, 4)
+    assert Limits(1, 2)*Limits(2, 3) == Limits(2, 6)
+
+    assert Limits(1, 2)*0 == 0
+    assert Limits(1, oo)*0 == Limits(0, oo)
+    assert Limits(-oo, 1)*0 == Limits(-oo, 0)
+    assert Limits(-oo, oo)*0 == Limits(-oo, oo)
+
+    assert Limits(1, 2)*x == Mul(Limits(1, 2), x, evaluate=False)
+
+    assert Limits(0, 2)*oo == Limits(0, oo)
+    assert Limits(-2, 0)*oo == Limits(-oo, 0)
+    assert Limits(0, 2)*(-oo) == Limits(-oo, 0)
+    assert Limits(-2, 0)*(-oo) == Limits(0, oo)
+    assert Limits(-1, 1)*oo == Limits(-oo, oo)
+    assert Limits(-1, 1)*(-oo) == Limits(-oo, oo)
+    assert Limits(-oo, oo) == Limits(-oo, oo)
+
+
+def test_Limits_div():
+    assert Limits(-1, 3)/Limits(3, 4) == Limits(-S(1)/3, 1)
+    assert Limits(-2, 4)/Limits(-3, 4) == Limits(-oo, oo)
+    assert Limits(-3, -2)/Limits(-4, 0) == Limits(S(1)/2, oo)
+#    assert Limits(-3, -2)/Limits(-2, 1) == Union(Limits(-oo, -2), Limits(1, oo))
+    assert Limits(-3, -2)/Limits(0, 4) == Limits(-oo, -S(1)/2)
+    assert Limits(2, 4)/Limits(-3, 0) == Limits(-oo, -S(2)/3)
+#    assert Limits(2, 3)/Limits(-2, 2) == Union(Limits(-oo, -1), Limits(1, oo))
+    assert Limits(2, 4)/Limits(0, 3) == Limits(S(2)/3, oo)
+
+    assert Limits(0, 1)/Limits(0, 1) == Limits(0, oo)
+    assert Limits(-1, 0)/Limits(0, 1) == Limits(-oo, 0)
+    assert Limits(-1, 2)/Limits(-2, 2) == Limits(-oo, oo)
+
+    assert 1/Limits(-1, 2) == Limits(-oo, oo)
+    assert 1/Limits(0, 2) == Limits(S(1)/2, oo)
+    assert (-1)/Limits(0, 2) == Limits(-oo, -S(1)/2)
+
+    assert Limits(1, 2)/a == Mul(Limits(1, 2), 1/a, evaluate=False)
+
+    assert Limits(1, 2)/0 == Limits(1, 2)*zoo
+    assert Limits(1, oo)/oo == Limits(0, oo)
+    assert Limits(1, oo)/(-oo) == Limits(-oo, 0)
+    assert Limits(-oo, -1)/oo == Limits(-oo, 0)
+    assert Limits(-oo, -1)/(-oo) == Limits(0, oo)
+    assert Limits(-oo, oo)/oo == Limits(-oo, oo)
+    assert Limits(-oo, oo)/(-oo) == Limits(-oo, oo)
+    assert Limits(-1, oo)/oo == Limits(0, oo)
+    assert Limits(-1, oo)/(-oo) == Limits(-oo, 0)
+    assert Limits(-oo, 1)/oo == Limits(-oo, 0)
+    assert Limits(-oo, 1)/(-oo) == Limits(0, oo)
+
+
+def test_Limits_subs():
+    assert (x**2 + 2*x + 1).subs(x, Limits(-1, 1)) == Limits(-1, 4)
```

Comment:
This should probably be inspected more closely. ``x**2 + 2*x + 1`` is non-negative for all real values of ``x``.